### PR TITLE
Accept an initial value

### LINKED ISSUE #4: Accept an initial value
Consider supporting the injection of an initial value into the loop. The code could be updated as follows:

```js
async function pDoWhilst(action, condition, initialValue) {
  const actionResult = await action(await initialValue);

  if (await condition(actionResult)) {
    return pDoWhilst(action, condition, actionResult);
  }

  return actionResult;
}
```

and that would allow managing the state internally,  instead of having to rely in an external variable:

```js
const finalCount = await pDoWhilst(
  (count) => count + 1,
  (count) => count < 5,
  0
);

console.log(finalCount);
//=> 5
```

Note that this change would be compatible and will not break the current API.

### PR DESCRIPTION
**Pull request by gabmontes**
The initial value will be passed to the action. The result will be passed to the next call to the action. The last result will be returned.

Tests, types and docs were updated.

Closes #4.

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -20,6 +20,7 @@ console.log(count);
 ```
 */
 export default function pDoWhilst<ValueType>(
```
Comment (sindresorhus) [inline index.d.ts:22]: You also need to update the doc comment.

File: readme.md [hunk 2]
```diff
@@ -43,8 +44,9 @@ You can return a promise and it will be handled.
 #### condition
 
 Type: `Function`
+Arguments: The value the `action` function returns
```
Comment (sindresorhus) [inline readme.md]:
```suggestion
Arguments: The value the `action` function returns.
```

File: index.js [hunk 3]
```diff
@@ -1,7 +1,9 @@
-export default async function pDoWhilst(action, condition) {
-	const actionResult = await action();
+export default async function pDoWhilst(action, condition, initialValue) {
+	const actionResult = await action(await initialValue);
```
Comment (sindresorhus) [inline index.js]: `initialValue` should not await the promise. That makes it impossible for the `initialValue` to be a promise you actually want passed into `action`, if you want to deal with the promise there.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: index.js (hunk 3 above)
Comment (gabmontes) [reply index.js]: You are right! If `initialValue` rejects for some reason, the exception will be thrown inside `pDoWhilst` preventing `action` do anything about it.

Changed in 2f44ae0. A test was added too.

File: readme.md (hunk 2 above)
Comment (gabmontes) [reply readme.md]: Fixed in 2f44ae0.

File: index.d.ts (hunk 1 above)
Comment (gabmontes) [reply index.d.ts:22]: Updated in 2f44ae0.

File: index.d.ts [hunk 4]
```diff
@@ -20,6 +20,7 @@ console.log(count);
...
-	action: () => ValueType | PromiseLike<ValueType>,
-	condition: (value: ValueType) => boolean
-): Promise<void>;
+	action: (value: ValueType | PromiseLike<ValueType>) => ValueType | PromiseLike<ValueType>,
+	condition: (value: ValueType | PromiseLike<ValueType>) => boolean | PromiseLike<boolean>,
+	initialValue?: ValueType | PromiseLike<ValueType>,
```
Comment (sindresorhus) [inline index.d.ts]: This is incorrect. It should preserve the input `ValueType`.

```suggestion
	action: (value: ValueType) => ValueType | PromiseLike<ValueType>,
	condition: (value: ValueType) => boolean | PromiseLike<boolean>,
	initialValue?: ValueType,
```

Comment (gabmontes) [reply index.d.ts]: Right! Fixed in cebefc8. Also rolled back the changes in `index.test-d.ts` as those were not actually needed.

**Comment by gabmontes** [comment]
@sindresorhus all findings addressed.

Re unresolved conversations, some devs are used to close them as soon as they address the issues. Others wait for the author of the conversation to close it as it is the "owner" of that conversation. What is the policy for this repo?

**Comment by sindresorhus** [comment]
> Others wait for the author of the conversation to close it as it is the "owner" of that conversation. What is the policy for this repo?

I think that's generally the best way to handle it, because the owner likely want to verify before marking it as resolved.

**Comment by sindresorhus** [comment]
Thanks :)

**Comment by gabmontes** [comment]
@sindresorhus thanks for creating these awesome packages in the first place and for the support to make this PR right!

**Pull request merged**